### ISSUE TITLE
`valid?` doesn't work with double-nulls

### ISSUE THREAD

**Issue by eapache** [rmosolgo/graphql-ruby#951]
Constructing a `GraphQL::Query` with a double-non-null variable like
```graphql
query getProduct($pid: String!!) { node(id: $pid) { id } }
```
and then calling `valid?` on it results in an exception:
```
GraphQL::DoubleNonNullTypeError: You tried to add a non-null constraint twice (!! instead of !)
```
instead of a simple `false`.

I'm assuming this is unintentional? cc @xuorig

**Comment by xuorig** [comment]
Ah it looks like this is raising when parsing => turning into an AST, even before validation

```ruby
    def initialize(of_type:)
      if of_type.is_a?(GraphQL::NonNullType)
        raise(
          DoubleNonNullTypeError,
          "You tried to add a non-null constraint twice (!! instead of !)"
        )
      end

      super()
      @of_type = of_type
    end
```

**Comment by xuorig** [comment]
but... isn't this invalid to the parser?

**Comment by eapache** [comment]
I guess the parsing is lazy and that's why the exception is coming from `valid?` and not the constructor?

**Comment by rmosolgo** [comment]
> parsing is lazy

yes! 

>  invalid to the parser?

It would be nice to catch it there, if we can make it happen ... 

Definitely should be fixed somehow to return an error to the client instead of :boom: ! 

(This error is meant to prevent double non-null types with ruby, eg `ThingType.to_non_null_type.to_non_null_type`)

**Comment by eapache** [comment]
*just* ran into this again

**Comment by rmosolgo** [comment]
sorry, tagged as a reminder!

**Comment by rmosolgo** [comment]
This is fixed on 1.10-dev and will be released in 1.10.0
